sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -429,6 +431,62 @@ private HttpPipeline setupPipeline(HttpClient httpClient) {
         }).map(ar -> new MsalToken(ar, options));
     }
 
+    /**
+     * Asynchronously acquire a token from Active Directory with a device code challenge. Active Directory will provide
+     * a device code for login and the user must meet the challenge by authenticating in a browser on the current or a
+     * different device.
+     *
+     * @param request the details of the token request
+     * @return a Publisher that emits an AccessToken when the device challenge is met, or an exception if the device
+     *     code expires
+     */
+    public Mono<MsalToken> authenticateWithVsCodeCredential(TokenRequestContext request) {
+
+        VisualStudioCacheAccessor accessor = new VisualStudioCacheAccessor();
+
+        JsonNode userSettings = accessor.getUserSettings();
+
+        String tenant = tenantId;
+        String cloud = "Azure";
+
+        if (!userSettings.isNull()) {
+            if (userSettings.has("azure.tenant")) {
+                tenant = userSettings.get("azure.tenant").asText();
+            }
+
+            if (userSettings.has("azure.cloud")) {
+                cloud = userSettings.get("azure.cloud").asText();
+            }
+        }
+
+        String credential = accessor.getCredentials("VS Code Azure", cloud);
+
+        String authority = accessor.getAzureAuthHost(cloud).replaceAll("/+$", "")
+            + "/organizations/" + tenant;
+        PublicClientApplication.Builder publicClientApplicationBuilder = PublicClientApplication.builder(clientId);
+        if (httpPipelineAdapter != null) {
+            publicClientApplicationBuilder.httpClient(httpPipelineAdapter);
+        } else if (options.getProxyOptions() != null) {
+            publicClientApplicationBuilder.proxy(proxyOptionsToJavaNetProxy(options.getProxyOptions()));
+        }
+
+        try {
+            PublicClientApplication clientApplication =  publicClientApplicationBuilder
+                                                             .authority(authority)
+                                                             .build();
+
+            RefreshTokenParameters parameters = RefreshTokenParameters
+                                                    .builder(new HashSet<>(request.getScopes()), credential)
+                                                    .build();
+
+            return Mono.defer(() -> Mono.fromFuture(clientApplication.acquireToken(parameters))
+                                        .map(ar -> new MsalToken(ar, options)));
+
+        } catch (MalformedURLException e) {
+            return Mono.error(logger.logExceptionAsError(new IllegalStateException(e)));
```

Comment:
Why did you choose to wrap the MalformedURLException in an IllegalStateException?